# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [2]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [3]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [4]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    
    source_sentences = [sentence for sentence in source_text.split('\n')]
    target_sentences = [sentence + ' <EOS>' for sentence in target_text.split('\n')]
    
    source_id_text = [[source_vocab_to_int[word] for word in sentence.split()] for sentence in source_sentences]
    target_id_text = [[target_vocab_to_int[word] for word in sentence.split()] for sentence in target_sentences]
    
    return source_id_text, target_id_text


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [7]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [8]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0


/usr/local/lib/python3.5/site-packages/ipykernel_launcher.py:15: UserWarning: No GPU found. Please use a GPU to train your neural network.
  from ipykernel import kernelapp as app


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [9]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    inputs = tf.placeholder(tf.int32, shape=[None,None], name= "input")
    targets = tf.placeholder(tf.int32, shape=[None,None], name= "targets")
    lrate = tf.placeholder(tf.float32, name= "learning_rate")
    keep_prob = tf.placeholder(tf.float32, name= "keep_prob")
    target_seq_lenth = tf.placeholder(tf.int32, shape=[None], name= "target_sequence_length")
    max_target_len = tf.reduce_max(target_seq_lenth, name= 'max_target_len')
    source_seq_length = tf.placeholder(tf.int32, shape=[None], name= "source_sequence_length")
    return inputs, targets, lrate, keep_prob, target_seq_lenth, max_target_len, source_seq_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [10]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    go_id = tf.constant(target_vocab_to_int['<GO>'], shape=(batch_size,1), dtype=tf.int32)
    processed_input = tf.concat([go_id,target_data[:,:-1]],1)
    return processed_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [11]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    def build_cell(rnn_size, keep_prob):
        lstm = tf.contrib.rnn.LSTMCell(rnn_size)
        lstm_drop = tf.contrib.rnn.DropoutWrapper(lstm, output_keep_prob=keep_prob)
        return lstm_drop
    # Stack them all
    stacked_lstm = tf.contrib.rnn.MultiRNNCell([build_cell(rnn_size, keep_prob) for _ in range(num_layers)])
    # Creathe embedding layer.
    embed_encoder = tf.contrib.layers.embed_sequence(rnn_inputs, vocab_size = source_vocab_size, embed_dim = encoding_embedding_size)
    # If we don't have an initial zero state, provide a dtype.
    output, state = tf.nn.dynamic_rnn(stacked_lstm, embed_encoder, source_sequence_length, dtype=tf.float32)
    return output, state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [12]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    trainig_helper = tf.contrib.seq2seq.TrainingHelper(dec_embed_input, target_sequence_length)
    basic_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, trainig_helper, encoder_state, output_layer)
    f_output, _ = tf.contrib.seq2seq.dynamic_decode(basic_decoder,maximum_iterations=max_summary_length)
    return f_output

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [13]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    start_ids = tf.tile([start_of_sequence_id], [batch_size])
    embedding_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(
        dec_embeddings, start_ids, end_of_sequence_id)
    basic_decoder = tf.contrib.seq2seq.BasicDecoder(
        dec_cell, embedding_helper, encoder_state, output_layer)
    f_output, _= tf.contrib.seq2seq.dynamic_decode(
        basic_decoder,maximum_iterations=max_target_sequence_length)
    return f_output

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [14]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    def build_cell(rnn_size, keep_prob):
        lstm = tf.contrib.rnn.LSTMCell(rnn_size)
        lstm_drop = tf.contrib.rnn.DropoutWrapper(lstm, output_keep_prob=keep_prob)
        return lstm_drop

    stacked_lstm = tf.contrib.rnn.MultiRNNCell([build_cell(rnn_size, keep_prob) for _ in range(num_layers)])
    
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)

    dense_layer = Dense(target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))
    
    with tf.variable_scope("decode") as scope:
        tr_decoder_output = decoding_layer_train(
            encoder_state, stacked_lstm, dec_embed_input, 
            target_sequence_length, max_target_sequence_length, 
            dense_layer, keep_prob)
        scope.reuse_variables()
        inf_decoder_output = decoding_layer_infer(
            encoder_state, stacked_lstm, dec_embeddings, 
            target_vocab_to_int['<GO>'], target_vocab_to_int['<EOS>'], 
            max_target_sequence_length, target_vocab_size, 
            dense_layer, batch_size, keep_prob)
    
    return tr_decoder_output, inf_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [15]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    output, state = encoding_layer(input_data, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   enc_embedding_size)
    
    processed_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    
    tr_decoder_output, inf_decoder_output = decoding_layer(processed_input, state,
                   target_sequence_length, max_target_sentence_length,
                   rnn_size, num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, dec_embedding_size)
    
    return tr_decoder_output, inf_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [16]:
# Number of Epochs
epochs = 15
# Batch Size
batch_size = 512
# RNN Size
rnn_size = 150
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 160
decoding_embedding_size = 160
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.5
display_step = True

### Build the Graph
Build the graph using the neural network you implemented.

In [17]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [18]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [19]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    1/269 - Train Accuracy: 0.2329, Validation Accuracy: 0.3096, Loss: 5.6265
Epoch   0 Batch    2/269 - Train Accuracy: 0.2655, Validation Accuracy: 0.3096, Loss: 5.3272
Epoch   0 Batch    3/269 - Train Accuracy: 0.2444, Validation Accuracy: 0.3096, Loss: 5.1177
Epoch   0 Batch    4/269 - Train Accuracy: 0.2317, Validation Accuracy: 0.3096, Loss: 4.9279
Epoch   0 Batch    5/269 - Train Accuracy: 0.2325, Validation Accuracy: 0.3096, Loss: 4.7660
Epoch   0 Batch    6/269 - Train Accuracy: 0.2786, Validation Accuracy: 0.3098, Loss: 4.4419
Epoch   0 Batch    7/269 - Train Accuracy: 0.2873, Validation Accuracy: 0.3220, Loss: 4.3754
Epoch   0 Batch    8/269 - Train Accuracy: 0.2659, Validation Accuracy: 0.3319, Loss: 4.4890
Epoch   0 Batch    9/269 - Train Accuracy: 0.2997, Validation Accuracy: 0.3410, Loss: 4.2399
Epoch   0 Batch   10/269 - Train Accuracy: 0.2692, Validation Accuracy: 0.3421, Loss: 4.2830
Epoch   0 Batch   11/269 - Train Accuracy: 0.3075, Validation Accuracy

Epoch   0 Batch   90/269 - Train Accuracy: 0.4175, Validation Accuracy: 0.4843, Loss: 2.4116
Epoch   0 Batch   91/269 - Train Accuracy: 0.4558, Validation Accuracy: 0.4847, Loss: 2.2580
Epoch   0 Batch   92/269 - Train Accuracy: 0.4556, Validation Accuracy: 0.4800, Loss: 2.2522
Epoch   0 Batch   93/269 - Train Accuracy: 0.4822, Validation Accuracy: 0.4850, Loss: 2.1721
Epoch   0 Batch   94/269 - Train Accuracy: 0.4575, Validation Accuracy: 0.4847, Loss: 2.2444
Epoch   0 Batch   95/269 - Train Accuracy: 0.4661, Validation Accuracy: 0.4877, Loss: 2.2227
Epoch   0 Batch   96/269 - Train Accuracy: 0.4609, Validation Accuracy: 0.4863, Loss: 2.2007
Epoch   0 Batch   97/269 - Train Accuracy: 0.4609, Validation Accuracy: 0.4889, Loss: 2.2098
Epoch   0 Batch   98/269 - Train Accuracy: 0.4596, Validation Accuracy: 0.4789, Loss: 2.1613
Epoch   0 Batch   99/269 - Train Accuracy: 0.4290, Validation Accuracy: 0.4828, Loss: 2.3159
Epoch   0 Batch  100/269 - Train Accuracy: 0.4750, Validation Accuracy

Epoch   0 Batch  179/269 - Train Accuracy: 0.4746, Validation Accuracy: 0.4972, Loss: 1.5430
Epoch   0 Batch  180/269 - Train Accuracy: 0.4763, Validation Accuracy: 0.5012, Loss: 1.5412
Epoch   0 Batch  181/269 - Train Accuracy: 0.4607, Validation Accuracy: 0.4814, Loss: 1.5481
Epoch   0 Batch  182/269 - Train Accuracy: 0.4566, Validation Accuracy: 0.4810, Loss: 1.5613
Epoch   0 Batch  183/269 - Train Accuracy: 0.5255, Validation Accuracy: 0.4898, Loss: 1.3353
Epoch   0 Batch  184/269 - Train Accuracy: 0.4250, Validation Accuracy: 0.4760, Loss: 1.5892
Epoch   0 Batch  185/269 - Train Accuracy: 0.4641, Validation Accuracy: 0.4790, Loss: 1.5149
Epoch   0 Batch  186/269 - Train Accuracy: 0.4432, Validation Accuracy: 0.4949, Loss: 1.6002
Epoch   0 Batch  187/269 - Train Accuracy: 0.4623, Validation Accuracy: 0.4854, Loss: 1.4943
Epoch   0 Batch  188/269 - Train Accuracy: 0.4621, Validation Accuracy: 0.4776, Loss: 1.4716
Epoch   0 Batch  189/269 - Train Accuracy: 0.4783, Validation Accuracy

Epoch   1 Batch    1/269 - Train Accuracy: 0.4829, Validation Accuracy: 0.5336, Loss: 1.2435
Epoch   1 Batch    2/269 - Train Accuracy: 0.4953, Validation Accuracy: 0.5344, Loss: 1.2150
Epoch   1 Batch    3/269 - Train Accuracy: 0.4938, Validation Accuracy: 0.5361, Loss: 1.2472
Epoch   1 Batch    4/269 - Train Accuracy: 0.4957, Validation Accuracy: 0.5375, Loss: 1.2369
Epoch   1 Batch    5/269 - Train Accuracy: 0.4823, Validation Accuracy: 0.5376, Loss: 1.2404
Epoch   1 Batch    6/269 - Train Accuracy: 0.5308, Validation Accuracy: 0.5402, Loss: 1.1360
Epoch   1 Batch    7/269 - Train Accuracy: 0.5180, Validation Accuracy: 0.5402, Loss: 1.1756
Epoch   1 Batch    8/269 - Train Accuracy: 0.5006, Validation Accuracy: 0.5390, Loss: 1.2371
Epoch   1 Batch    9/269 - Train Accuracy: 0.5010, Validation Accuracy: 0.5400, Loss: 1.1832
Epoch   1 Batch   10/269 - Train Accuracy: 0.4979, Validation Accuracy: 0.5412, Loss: 1.2041
Epoch   1 Batch   11/269 - Train Accuracy: 0.5117, Validation Accuracy

Epoch   1 Batch   90/269 - Train Accuracy: 0.4973, Validation Accuracy: 0.5596, Loss: 1.0357
Epoch   1 Batch   91/269 - Train Accuracy: 0.5308, Validation Accuracy: 0.5597, Loss: 0.9594
Epoch   1 Batch   92/269 - Train Accuracy: 0.5405, Validation Accuracy: 0.5593, Loss: 0.9586
Epoch   1 Batch   93/269 - Train Accuracy: 0.5614, Validation Accuracy: 0.5625, Loss: 0.9276
Epoch   1 Batch   94/269 - Train Accuracy: 0.5360, Validation Accuracy: 0.5633, Loss: 0.9919
Epoch   1 Batch   95/269 - Train Accuracy: 0.5520, Validation Accuracy: 0.5637, Loss: 0.9750
Epoch   1 Batch   96/269 - Train Accuracy: 0.5449, Validation Accuracy: 0.5614, Loss: 0.9645
Epoch   1 Batch   97/269 - Train Accuracy: 0.5303, Validation Accuracy: 0.5618, Loss: 0.9609
Epoch   1 Batch   98/269 - Train Accuracy: 0.5580, Validation Accuracy: 0.5588, Loss: 0.9516
Epoch   1 Batch   99/269 - Train Accuracy: 0.5223, Validation Accuracy: 0.5581, Loss: 1.0009
Epoch   1 Batch  100/269 - Train Accuracy: 0.5552, Validation Accuracy

Epoch   1 Batch  179/269 - Train Accuracy: 0.6024, Validation Accuracy: 0.5938, Loss: 0.8526
Epoch   1 Batch  180/269 - Train Accuracy: 0.5885, Validation Accuracy: 0.5933, Loss: 0.8284
Epoch   1 Batch  181/269 - Train Accuracy: 0.5779, Validation Accuracy: 0.5938, Loss: 0.8471
Epoch   1 Batch  182/269 - Train Accuracy: 0.5959, Validation Accuracy: 0.5936, Loss: 0.8403
Epoch   1 Batch  183/269 - Train Accuracy: 0.6430, Validation Accuracy: 0.5930, Loss: 0.7259
Epoch   1 Batch  184/269 - Train Accuracy: 0.5686, Validation Accuracy: 0.5936, Loss: 0.8707
Epoch   1 Batch  185/269 - Train Accuracy: 0.6007, Validation Accuracy: 0.5969, Loss: 0.8273
Epoch   1 Batch  186/269 - Train Accuracy: 0.5650, Validation Accuracy: 0.5982, Loss: 0.8541
Epoch   1 Batch  187/269 - Train Accuracy: 0.5966, Validation Accuracy: 0.5963, Loss: 0.8226
Epoch   1 Batch  188/269 - Train Accuracy: 0.5977, Validation Accuracy: 0.5956, Loss: 0.8127
Epoch   1 Batch  189/269 - Train Accuracy: 0.5896, Validation Accuracy

Epoch   2 Batch    1/269 - Train Accuracy: 0.5970, Validation Accuracy: 0.6159, Loss: 0.7620
Epoch   2 Batch    2/269 - Train Accuracy: 0.5881, Validation Accuracy: 0.6172, Loss: 0.7489
Epoch   2 Batch    3/269 - Train Accuracy: 0.6068, Validation Accuracy: 0.6173, Loss: 0.7664
Epoch   2 Batch    4/269 - Train Accuracy: 0.5825, Validation Accuracy: 0.6207, Loss: 0.7743
Epoch   2 Batch    5/269 - Train Accuracy: 0.5861, Validation Accuracy: 0.6183, Loss: 0.7775
Epoch   2 Batch    6/269 - Train Accuracy: 0.6189, Validation Accuracy: 0.6170, Loss: 0.7108
Epoch   2 Batch    7/269 - Train Accuracy: 0.6097, Validation Accuracy: 0.6217, Loss: 0.7284
Epoch   2 Batch    8/269 - Train Accuracy: 0.5886, Validation Accuracy: 0.6184, Loss: 0.7778
Epoch   2 Batch    9/269 - Train Accuracy: 0.6005, Validation Accuracy: 0.6204, Loss: 0.7488
Epoch   2 Batch   10/269 - Train Accuracy: 0.6031, Validation Accuracy: 0.6233, Loss: 0.7565
Epoch   2 Batch   11/269 - Train Accuracy: 0.5995, Validation Accuracy

Epoch   2 Batch   90/269 - Train Accuracy: 0.5908, Validation Accuracy: 0.6215, Loss: 0.7176
Epoch   2 Batch   91/269 - Train Accuracy: 0.6195, Validation Accuracy: 0.6246, Loss: 0.6542
Epoch   2 Batch   92/269 - Train Accuracy: 0.6187, Validation Accuracy: 0.6297, Loss: 0.6645
Epoch   2 Batch   93/269 - Train Accuracy: 0.6279, Validation Accuracy: 0.6281, Loss: 0.6424
Epoch   2 Batch   94/269 - Train Accuracy: 0.6220, Validation Accuracy: 0.6257, Loss: 0.6865
Epoch   2 Batch   95/269 - Train Accuracy: 0.6169, Validation Accuracy: 0.6257, Loss: 0.6760
Epoch   2 Batch   96/269 - Train Accuracy: 0.6232, Validation Accuracy: 0.6270, Loss: 0.6694
Epoch   2 Batch   97/269 - Train Accuracy: 0.6128, Validation Accuracy: 0.6311, Loss: 0.6681
Epoch   2 Batch   98/269 - Train Accuracy: 0.6364, Validation Accuracy: 0.6297, Loss: 0.6660
Epoch   2 Batch   99/269 - Train Accuracy: 0.6146, Validation Accuracy: 0.6335, Loss: 0.6870
Epoch   2 Batch  100/269 - Train Accuracy: 0.6421, Validation Accuracy

Epoch   2 Batch  179/269 - Train Accuracy: 0.6370, Validation Accuracy: 0.6323, Loss: 0.6244
Epoch   2 Batch  180/269 - Train Accuracy: 0.6469, Validation Accuracy: 0.6340, Loss: 0.6041
Epoch   2 Batch  181/269 - Train Accuracy: 0.6290, Validation Accuracy: 0.6331, Loss: 0.6085
Epoch   2 Batch  182/269 - Train Accuracy: 0.6426, Validation Accuracy: 0.6362, Loss: 0.6103
Epoch   2 Batch  183/269 - Train Accuracy: 0.6728, Validation Accuracy: 0.6289, Loss: 0.5313
Epoch   2 Batch  184/269 - Train Accuracy: 0.6098, Validation Accuracy: 0.6369, Loss: 0.6425
Epoch   2 Batch  185/269 - Train Accuracy: 0.6390, Validation Accuracy: 0.6392, Loss: 0.6052
Epoch   2 Batch  186/269 - Train Accuracy: 0.6224, Validation Accuracy: 0.6341, Loss: 0.6218
Epoch   2 Batch  187/269 - Train Accuracy: 0.6397, Validation Accuracy: 0.6384, Loss: 0.5998
Epoch   2 Batch  188/269 - Train Accuracy: 0.6509, Validation Accuracy: 0.6409, Loss: 0.5861
Epoch   2 Batch  189/269 - Train Accuracy: 0.6405, Validation Accuracy

Epoch   3 Batch    1/269 - Train Accuracy: 0.6507, Validation Accuracy: 0.6694, Loss: 0.5771
Epoch   3 Batch    2/269 - Train Accuracy: 0.6577, Validation Accuracy: 0.6650, Loss: 0.5674
Epoch   3 Batch    3/269 - Train Accuracy: 0.6521, Validation Accuracy: 0.6672, Loss: 0.5791
Epoch   3 Batch    4/269 - Train Accuracy: 0.6312, Validation Accuracy: 0.6624, Loss: 0.5908
Epoch   3 Batch    5/269 - Train Accuracy: 0.6405, Validation Accuracy: 0.6639, Loss: 0.5838
Epoch   3 Batch    6/269 - Train Accuracy: 0.6651, Validation Accuracy: 0.6622, Loss: 0.5464
Epoch   3 Batch    7/269 - Train Accuracy: 0.6555, Validation Accuracy: 0.6659, Loss: 0.5512
Epoch   3 Batch    8/269 - Train Accuracy: 0.6353, Validation Accuracy: 0.6590, Loss: 0.5899
Epoch   3 Batch    9/269 - Train Accuracy: 0.6548, Validation Accuracy: 0.6641, Loss: 0.5822
Epoch   3 Batch   10/269 - Train Accuracy: 0.6558, Validation Accuracy: 0.6681, Loss: 0.5787
Epoch   3 Batch   11/269 - Train Accuracy: 0.6391, Validation Accuracy

Epoch   3 Batch   90/269 - Train Accuracy: 0.6222, Validation Accuracy: 0.6707, Loss: 0.5557
Epoch   3 Batch   91/269 - Train Accuracy: 0.6870, Validation Accuracy: 0.6705, Loss: 0.5076
Epoch   3 Batch   92/269 - Train Accuracy: 0.6760, Validation Accuracy: 0.6689, Loss: 0.5131
Epoch   3 Batch   93/269 - Train Accuracy: 0.6830, Validation Accuracy: 0.6712, Loss: 0.4993
Epoch   3 Batch   94/269 - Train Accuracy: 0.6625, Validation Accuracy: 0.6709, Loss: 0.5315
Epoch   3 Batch   95/269 - Train Accuracy: 0.6584, Validation Accuracy: 0.6752, Loss: 0.5257
Epoch   3 Batch   96/269 - Train Accuracy: 0.6711, Validation Accuracy: 0.6747, Loss: 0.5266
Epoch   3 Batch   97/269 - Train Accuracy: 0.6749, Validation Accuracy: 0.6731, Loss: 0.5139
Epoch   3 Batch   98/269 - Train Accuracy: 0.6537, Validation Accuracy: 0.6692, Loss: 0.5220
Epoch   3 Batch   99/269 - Train Accuracy: 0.6544, Validation Accuracy: 0.6740, Loss: 0.5339
Epoch   3 Batch  100/269 - Train Accuracy: 0.6797, Validation Accuracy

Epoch   3 Batch  179/269 - Train Accuracy: 0.6831, Validation Accuracy: 0.6942, Loss: 0.4959
Epoch   3 Batch  180/269 - Train Accuracy: 0.6759, Validation Accuracy: 0.6875, Loss: 0.4825
Epoch   3 Batch  181/269 - Train Accuracy: 0.6738, Validation Accuracy: 0.6815, Loss: 0.4932
Epoch   3 Batch  182/269 - Train Accuracy: 0.6819, Validation Accuracy: 0.6802, Loss: 0.4959
Epoch   3 Batch  183/269 - Train Accuracy: 0.7209, Validation Accuracy: 0.6802, Loss: 0.4244
Epoch   3 Batch  184/269 - Train Accuracy: 0.6681, Validation Accuracy: 0.6803, Loss: 0.5035
Epoch   3 Batch  185/269 - Train Accuracy: 0.6882, Validation Accuracy: 0.6832, Loss: 0.4843
Epoch   3 Batch  186/269 - Train Accuracy: 0.6631, Validation Accuracy: 0.6729, Loss: 0.4924
Epoch   3 Batch  187/269 - Train Accuracy: 0.6932, Validation Accuracy: 0.6962, Loss: 0.4796
Epoch   3 Batch  188/269 - Train Accuracy: 0.6953, Validation Accuracy: 0.6984, Loss: 0.4706
Epoch   3 Batch  189/269 - Train Accuracy: 0.6671, Validation Accuracy

Epoch   4 Batch    1/269 - Train Accuracy: 0.7219, Validation Accuracy: 0.7291, Loss: 0.4643
Epoch   4 Batch    2/269 - Train Accuracy: 0.7143, Validation Accuracy: 0.7249, Loss: 0.4519
Epoch   4 Batch    3/269 - Train Accuracy: 0.7044, Validation Accuracy: 0.7093, Loss: 0.4528
Epoch   4 Batch    4/269 - Train Accuracy: 0.7069, Validation Accuracy: 0.7318, Loss: 0.4692
Epoch   4 Batch    5/269 - Train Accuracy: 0.7054, Validation Accuracy: 0.7270, Loss: 0.4703
Epoch   4 Batch    6/269 - Train Accuracy: 0.7565, Validation Accuracy: 0.7297, Loss: 0.4314
Epoch   4 Batch    7/269 - Train Accuracy: 0.7235, Validation Accuracy: 0.7300, Loss: 0.4371
Epoch   4 Batch    8/269 - Train Accuracy: 0.7091, Validation Accuracy: 0.7206, Loss: 0.4651
Epoch   4 Batch    9/269 - Train Accuracy: 0.7140, Validation Accuracy: 0.7228, Loss: 0.4732
Epoch   4 Batch   10/269 - Train Accuracy: 0.7271, Validation Accuracy: 0.7285, Loss: 0.4608
Epoch   4 Batch   11/269 - Train Accuracy: 0.7315, Validation Accuracy

Epoch   4 Batch   90/269 - Train Accuracy: 0.7250, Validation Accuracy: 0.7512, Loss: 0.4449
Epoch   4 Batch   91/269 - Train Accuracy: 0.7497, Validation Accuracy: 0.7452, Loss: 0.3972
Epoch   4 Batch   92/269 - Train Accuracy: 0.7291, Validation Accuracy: 0.7291, Loss: 0.4033
Epoch   4 Batch   93/269 - Train Accuracy: 0.7484, Validation Accuracy: 0.7417, Loss: 0.3925
Epoch   4 Batch   94/269 - Train Accuracy: 0.7381, Validation Accuracy: 0.7432, Loss: 0.4228
Epoch   4 Batch   95/269 - Train Accuracy: 0.7299, Validation Accuracy: 0.7424, Loss: 0.4108
Epoch   4 Batch   96/269 - Train Accuracy: 0.7330, Validation Accuracy: 0.7243, Loss: 0.4037
Epoch   4 Batch   97/269 - Train Accuracy: 0.7350, Validation Accuracy: 0.7393, Loss: 0.4137
Epoch   4 Batch   98/269 - Train Accuracy: 0.7375, Validation Accuracy: 0.7395, Loss: 0.4142
Epoch   4 Batch   99/269 - Train Accuracy: 0.7250, Validation Accuracy: 0.7338, Loss: 0.4210
Epoch   4 Batch  100/269 - Train Accuracy: 0.7573, Validation Accuracy

Epoch   4 Batch  179/269 - Train Accuracy: 0.7510, Validation Accuracy: 0.7481, Loss: 0.3834
Epoch   4 Batch  180/269 - Train Accuracy: 0.7626, Validation Accuracy: 0.7546, Loss: 0.3764
Epoch   4 Batch  181/269 - Train Accuracy: 0.7534, Validation Accuracy: 0.7531, Loss: 0.3803
Epoch   4 Batch  182/269 - Train Accuracy: 0.7679, Validation Accuracy: 0.7606, Loss: 0.3887
Epoch   4 Batch  183/269 - Train Accuracy: 0.7801, Validation Accuracy: 0.7282, Loss: 0.3266
Epoch   4 Batch  184/269 - Train Accuracy: 0.7377, Validation Accuracy: 0.7480, Loss: 0.4018
Epoch   4 Batch  185/269 - Train Accuracy: 0.7556, Validation Accuracy: 0.7591, Loss: 0.3826
Epoch   4 Batch  186/269 - Train Accuracy: 0.7311, Validation Accuracy: 0.7339, Loss: 0.3860
Epoch   4 Batch  187/269 - Train Accuracy: 0.7466, Validation Accuracy: 0.7498, Loss: 0.3812
Epoch   4 Batch  188/269 - Train Accuracy: 0.7512, Validation Accuracy: 0.7537, Loss: 0.3751
Epoch   4 Batch  189/269 - Train Accuracy: 0.7626, Validation Accuracy

Epoch   5 Batch    1/269 - Train Accuracy: 0.7751, Validation Accuracy: 0.7650, Loss: 0.3566
Epoch   5 Batch    2/269 - Train Accuracy: 0.7739, Validation Accuracy: 0.7801, Loss: 0.3526
Epoch   5 Batch    3/269 - Train Accuracy: 0.7712, Validation Accuracy: 0.7667, Loss: 0.3597
Epoch   5 Batch    4/269 - Train Accuracy: 0.7410, Validation Accuracy: 0.7584, Loss: 0.3814
Epoch   5 Batch    5/269 - Train Accuracy: 0.7536, Validation Accuracy: 0.7766, Loss: 0.3692
Epoch   5 Batch    6/269 - Train Accuracy: 0.7928, Validation Accuracy: 0.7652, Loss: 0.3410
Epoch   5 Batch    7/269 - Train Accuracy: 0.7754, Validation Accuracy: 0.7574, Loss: 0.3493
Epoch   5 Batch    8/269 - Train Accuracy: 0.7747, Validation Accuracy: 0.7824, Loss: 0.3685
Epoch   5 Batch    9/269 - Train Accuracy: 0.7725, Validation Accuracy: 0.7800, Loss: 0.3650
Epoch   5 Batch   10/269 - Train Accuracy: 0.7596, Validation Accuracy: 0.7723, Loss: 0.3497
Epoch   5 Batch   11/269 - Train Accuracy: 0.7861, Validation Accuracy

Epoch   5 Batch   90/269 - Train Accuracy: 0.7723, Validation Accuracy: 0.7850, Loss: 0.3389
Epoch   5 Batch   91/269 - Train Accuracy: 0.8057, Validation Accuracy: 0.7722, Loss: 0.3105
Epoch   5 Batch   92/269 - Train Accuracy: 0.8013, Validation Accuracy: 0.7817, Loss: 0.3111
Epoch   5 Batch   93/269 - Train Accuracy: 0.8017, Validation Accuracy: 0.7859, Loss: 0.2996
Epoch   5 Batch   94/269 - Train Accuracy: 0.7919, Validation Accuracy: 0.7807, Loss: 0.3320
Epoch   5 Batch   95/269 - Train Accuracy: 0.7841, Validation Accuracy: 0.7896, Loss: 0.3151
Epoch   5 Batch   96/269 - Train Accuracy: 0.7787, Validation Accuracy: 0.7832, Loss: 0.3152
Epoch   5 Batch   97/269 - Train Accuracy: 0.7915, Validation Accuracy: 0.7822, Loss: 0.3157
Epoch   5 Batch   98/269 - Train Accuracy: 0.7797, Validation Accuracy: 0.7817, Loss: 0.3189
Epoch   5 Batch   99/269 - Train Accuracy: 0.7891, Validation Accuracy: 0.7789, Loss: 0.3269
Epoch   5 Batch  100/269 - Train Accuracy: 0.8173, Validation Accuracy

Epoch   5 Batch  179/269 - Train Accuracy: 0.8021, Validation Accuracy: 0.8042, Loss: 0.2971
Epoch   5 Batch  180/269 - Train Accuracy: 0.8046, Validation Accuracy: 0.7874, Loss: 0.2876
Epoch   5 Batch  181/269 - Train Accuracy: 0.7985, Validation Accuracy: 0.8113, Loss: 0.3038
Epoch   5 Batch  182/269 - Train Accuracy: 0.8173, Validation Accuracy: 0.8119, Loss: 0.2915
Epoch   5 Batch  183/269 - Train Accuracy: 0.8280, Validation Accuracy: 0.7875, Loss: 0.2580
Epoch   5 Batch  184/269 - Train Accuracy: 0.7989, Validation Accuracy: 0.8026, Loss: 0.3065
Epoch   5 Batch  185/269 - Train Accuracy: 0.8120, Validation Accuracy: 0.7993, Loss: 0.2938
Epoch   5 Batch  186/269 - Train Accuracy: 0.8093, Validation Accuracy: 0.7953, Loss: 0.2964
Epoch   5 Batch  187/269 - Train Accuracy: 0.8016, Validation Accuracy: 0.8051, Loss: 0.2876
Epoch   5 Batch  188/269 - Train Accuracy: 0.8155, Validation Accuracy: 0.8059, Loss: 0.2758
Epoch   5 Batch  189/269 - Train Accuracy: 0.8220, Validation Accuracy

Epoch   6 Batch    1/269 - Train Accuracy: 0.8096, Validation Accuracy: 0.8195, Loss: 0.2769
Epoch   6 Batch    2/269 - Train Accuracy: 0.8125, Validation Accuracy: 0.8208, Loss: 0.2869
Epoch   6 Batch    3/269 - Train Accuracy: 0.8311, Validation Accuracy: 0.8132, Loss: 0.2709
Epoch   6 Batch    4/269 - Train Accuracy: 0.7901, Validation Accuracy: 0.8202, Loss: 0.2867
Epoch   6 Batch    5/269 - Train Accuracy: 0.7900, Validation Accuracy: 0.8193, Loss: 0.2833
Epoch   6 Batch    6/269 - Train Accuracy: 0.8321, Validation Accuracy: 0.8119, Loss: 0.2613
Epoch   6 Batch    7/269 - Train Accuracy: 0.8246, Validation Accuracy: 0.8101, Loss: 0.2632
Epoch   6 Batch    8/269 - Train Accuracy: 0.7999, Validation Accuracy: 0.7932, Loss: 0.2828
Epoch   6 Batch    9/269 - Train Accuracy: 0.8198, Validation Accuracy: 0.8077, Loss: 0.2908
Epoch   6 Batch   10/269 - Train Accuracy: 0.8240, Validation Accuracy: 0.8161, Loss: 0.2811
Epoch   6 Batch   11/269 - Train Accuracy: 0.8158, Validation Accuracy

Epoch   6 Batch   90/269 - Train Accuracy: 0.8144, Validation Accuracy: 0.8271, Loss: 0.2672
Epoch   6 Batch   91/269 - Train Accuracy: 0.8545, Validation Accuracy: 0.8404, Loss: 0.2377
Epoch   6 Batch   92/269 - Train Accuracy: 0.8509, Validation Accuracy: 0.8351, Loss: 0.2394
Epoch   6 Batch   93/269 - Train Accuracy: 0.8364, Validation Accuracy: 0.8278, Loss: 0.2300
Epoch   6 Batch   94/269 - Train Accuracy: 0.8364, Validation Accuracy: 0.8295, Loss: 0.2693
Epoch   6 Batch   95/269 - Train Accuracy: 0.8346, Validation Accuracy: 0.8366, Loss: 0.2471
Epoch   6 Batch   96/269 - Train Accuracy: 0.8171, Validation Accuracy: 0.8296, Loss: 0.2444
Epoch   6 Batch   97/269 - Train Accuracy: 0.8376, Validation Accuracy: 0.8184, Loss: 0.2432
Epoch   6 Batch   98/269 - Train Accuracy: 0.8294, Validation Accuracy: 0.8407, Loss: 0.2475
Epoch   6 Batch   99/269 - Train Accuracy: 0.8260, Validation Accuracy: 0.8367, Loss: 0.2519
Epoch   6 Batch  100/269 - Train Accuracy: 0.8514, Validation Accuracy

Epoch   6 Batch  179/269 - Train Accuracy: 0.8369, Validation Accuracy: 0.8384, Loss: 0.2260
Epoch   6 Batch  180/269 - Train Accuracy: 0.8436, Validation Accuracy: 0.8350, Loss: 0.2190
Epoch   6 Batch  181/269 - Train Accuracy: 0.8458, Validation Accuracy: 0.8431, Loss: 0.2338
Epoch   6 Batch  182/269 - Train Accuracy: 0.8491, Validation Accuracy: 0.8471, Loss: 0.2183
Epoch   6 Batch  183/269 - Train Accuracy: 0.8784, Validation Accuracy: 0.8423, Loss: 0.1977
Epoch   6 Batch  184/269 - Train Accuracy: 0.8371, Validation Accuracy: 0.8435, Loss: 0.2324
Epoch   6 Batch  185/269 - Train Accuracy: 0.8342, Validation Accuracy: 0.8344, Loss: 0.2190
Epoch   6 Batch  186/269 - Train Accuracy: 0.8507, Validation Accuracy: 0.8390, Loss: 0.2262
Epoch   6 Batch  187/269 - Train Accuracy: 0.8528, Validation Accuracy: 0.8384, Loss: 0.2227
Epoch   6 Batch  188/269 - Train Accuracy: 0.8542, Validation Accuracy: 0.8454, Loss: 0.2157
Epoch   6 Batch  189/269 - Train Accuracy: 0.8642, Validation Accuracy

Epoch   7 Batch    1/269 - Train Accuracy: 0.8665, Validation Accuracy: 0.8621, Loss: 0.2153
Epoch   7 Batch    2/269 - Train Accuracy: 0.8543, Validation Accuracy: 0.8595, Loss: 0.2180
Epoch   7 Batch    3/269 - Train Accuracy: 0.8717, Validation Accuracy: 0.8498, Loss: 0.2133
Epoch   7 Batch    4/269 - Train Accuracy: 0.8360, Validation Accuracy: 0.8531, Loss: 0.2223
Epoch   7 Batch    5/269 - Train Accuracy: 0.8389, Validation Accuracy: 0.8516, Loss: 0.2137
Epoch   7 Batch    6/269 - Train Accuracy: 0.8758, Validation Accuracy: 0.8526, Loss: 0.1994
Epoch   7 Batch    7/269 - Train Accuracy: 0.8607, Validation Accuracy: 0.8532, Loss: 0.2061
Epoch   7 Batch    8/269 - Train Accuracy: 0.8673, Validation Accuracy: 0.8507, Loss: 0.2213
Epoch   7 Batch    9/269 - Train Accuracy: 0.8687, Validation Accuracy: 0.8532, Loss: 0.2196
Epoch   7 Batch   10/269 - Train Accuracy: 0.8553, Validation Accuracy: 0.8648, Loss: 0.2059
Epoch   7 Batch   11/269 - Train Accuracy: 0.8610, Validation Accuracy

Epoch   7 Batch   90/269 - Train Accuracy: 0.8568, Validation Accuracy: 0.8623, Loss: 0.2118
Epoch   7 Batch   91/269 - Train Accuracy: 0.8797, Validation Accuracy: 0.8614, Loss: 0.1780
Epoch   7 Batch   92/269 - Train Accuracy: 0.8847, Validation Accuracy: 0.8619, Loss: 0.1889
Epoch   7 Batch   93/269 - Train Accuracy: 0.8691, Validation Accuracy: 0.8589, Loss: 0.1857
Epoch   7 Batch   94/269 - Train Accuracy: 0.8653, Validation Accuracy: 0.8619, Loss: 0.2150
Epoch   7 Batch   95/269 - Train Accuracy: 0.8756, Validation Accuracy: 0.8619, Loss: 0.1925
Epoch   7 Batch   96/269 - Train Accuracy: 0.8567, Validation Accuracy: 0.8647, Loss: 0.1939
Epoch   7 Batch   97/269 - Train Accuracy: 0.8748, Validation Accuracy: 0.8624, Loss: 0.1974
Epoch   7 Batch   98/269 - Train Accuracy: 0.8711, Validation Accuracy: 0.8646, Loss: 0.1965
Epoch   7 Batch   99/269 - Train Accuracy: 0.8673, Validation Accuracy: 0.8678, Loss: 0.2006
Epoch   7 Batch  100/269 - Train Accuracy: 0.8923, Validation Accuracy

Epoch   7 Batch  179/269 - Train Accuracy: 0.8664, Validation Accuracy: 0.8653, Loss: 0.1854
Epoch   7 Batch  180/269 - Train Accuracy: 0.8762, Validation Accuracy: 0.8719, Loss: 0.1816
Epoch   7 Batch  181/269 - Train Accuracy: 0.8741, Validation Accuracy: 0.8722, Loss: 0.1883
Epoch   7 Batch  182/269 - Train Accuracy: 0.8786, Validation Accuracy: 0.8724, Loss: 0.1862
Epoch   7 Batch  183/269 - Train Accuracy: 0.8953, Validation Accuracy: 0.8675, Loss: 0.1556
Epoch   7 Batch  184/269 - Train Accuracy: 0.8782, Validation Accuracy: 0.8677, Loss: 0.1867
Epoch   7 Batch  185/269 - Train Accuracy: 0.8834, Validation Accuracy: 0.8705, Loss: 0.1755
Epoch   7 Batch  186/269 - Train Accuracy: 0.8768, Validation Accuracy: 0.8732, Loss: 0.1694
Epoch   7 Batch  187/269 - Train Accuracy: 0.8788, Validation Accuracy: 0.8712, Loss: 0.1701
Epoch   7 Batch  188/269 - Train Accuracy: 0.8868, Validation Accuracy: 0.8714, Loss: 0.1754
Epoch   7 Batch  189/269 - Train Accuracy: 0.8993, Validation Accuracy

Epoch   8 Batch    1/269 - Train Accuracy: 0.8902, Validation Accuracy: 0.8816, Loss: 0.1688
Epoch   8 Batch    2/269 - Train Accuracy: 0.8815, Validation Accuracy: 0.8809, Loss: 0.1703
Epoch   8 Batch    3/269 - Train Accuracy: 0.9024, Validation Accuracy: 0.8875, Loss: 0.1642
Epoch   8 Batch    4/269 - Train Accuracy: 0.8731, Validation Accuracy: 0.8810, Loss: 0.1695
Epoch   8 Batch    5/269 - Train Accuracy: 0.8706, Validation Accuracy: 0.8751, Loss: 0.1683
Epoch   8 Batch    6/269 - Train Accuracy: 0.9088, Validation Accuracy: 0.8857, Loss: 0.1566
Epoch   8 Batch    7/269 - Train Accuracy: 0.8850, Validation Accuracy: 0.8901, Loss: 0.1603
Epoch   8 Batch    8/269 - Train Accuracy: 0.8978, Validation Accuracy: 0.8868, Loss: 0.1768
Epoch   8 Batch    9/269 - Train Accuracy: 0.8890, Validation Accuracy: 0.8804, Loss: 0.1725
Epoch   8 Batch   10/269 - Train Accuracy: 0.8911, Validation Accuracy: 0.8755, Loss: 0.1637
Epoch   8 Batch   11/269 - Train Accuracy: 0.8879, Validation Accuracy

Epoch   8 Batch   90/269 - Train Accuracy: 0.8918, Validation Accuracy: 0.8846, Loss: 0.1650
Epoch   8 Batch   91/269 - Train Accuracy: 0.9044, Validation Accuracy: 0.8916, Loss: 0.1435
Epoch   8 Batch   92/269 - Train Accuracy: 0.9093, Validation Accuracy: 0.8920, Loss: 0.1460
Epoch   8 Batch   93/269 - Train Accuracy: 0.8993, Validation Accuracy: 0.8921, Loss: 0.1421
Epoch   8 Batch   94/269 - Train Accuracy: 0.8841, Validation Accuracy: 0.8955, Loss: 0.1591
Epoch   8 Batch   95/269 - Train Accuracy: 0.8929, Validation Accuracy: 0.8933, Loss: 0.1490
Epoch   8 Batch   96/269 - Train Accuracy: 0.8925, Validation Accuracy: 0.8971, Loss: 0.1477
Epoch   8 Batch   97/269 - Train Accuracy: 0.8983, Validation Accuracy: 0.8873, Loss: 0.1586
Epoch   8 Batch   98/269 - Train Accuracy: 0.8976, Validation Accuracy: 0.8905, Loss: 0.1527
Epoch   8 Batch   99/269 - Train Accuracy: 0.8947, Validation Accuracy: 0.8912, Loss: 0.1533
Epoch   8 Batch  100/269 - Train Accuracy: 0.9088, Validation Accuracy

Epoch   8 Batch  179/269 - Train Accuracy: 0.8828, Validation Accuracy: 0.9001, Loss: 0.1400
Epoch   8 Batch  180/269 - Train Accuracy: 0.9056, Validation Accuracy: 0.8966, Loss: 0.1369
Epoch   8 Batch  181/269 - Train Accuracy: 0.8890, Validation Accuracy: 0.9031, Loss: 0.1438
Epoch   8 Batch  182/269 - Train Accuracy: 0.8935, Validation Accuracy: 0.9017, Loss: 0.1410
Epoch   8 Batch  183/269 - Train Accuracy: 0.9112, Validation Accuracy: 0.8999, Loss: 0.1185
Epoch   8 Batch  184/269 - Train Accuracy: 0.9099, Validation Accuracy: 0.9000, Loss: 0.1406
Epoch   8 Batch  185/269 - Train Accuracy: 0.9041, Validation Accuracy: 0.8957, Loss: 0.1391
Epoch   8 Batch  186/269 - Train Accuracy: 0.8991, Validation Accuracy: 0.8995, Loss: 0.1411
Epoch   8 Batch  187/269 - Train Accuracy: 0.9018, Validation Accuracy: 0.8971, Loss: 0.1377
Epoch   8 Batch  188/269 - Train Accuracy: 0.9053, Validation Accuracy: 0.8925, Loss: 0.1346
Epoch   8 Batch  189/269 - Train Accuracy: 0.9067, Validation Accuracy

Epoch   9 Batch    1/269 - Train Accuracy: 0.9037, Validation Accuracy: 0.8979, Loss: 0.1317
Epoch   9 Batch    2/269 - Train Accuracy: 0.8920, Validation Accuracy: 0.8975, Loss: 0.1360
Epoch   9 Batch    3/269 - Train Accuracy: 0.9117, Validation Accuracy: 0.8967, Loss: 0.1399
Epoch   9 Batch    4/269 - Train Accuracy: 0.8948, Validation Accuracy: 0.9011, Loss: 0.1443
Epoch   9 Batch    5/269 - Train Accuracy: 0.8831, Validation Accuracy: 0.8948, Loss: 0.1282
Epoch   9 Batch    6/269 - Train Accuracy: 0.9180, Validation Accuracy: 0.8960, Loss: 0.1205
Epoch   9 Batch    7/269 - Train Accuracy: 0.8989, Validation Accuracy: 0.9023, Loss: 0.1293
Epoch   9 Batch    8/269 - Train Accuracy: 0.9071, Validation Accuracy: 0.9031, Loss: 0.1338
Epoch   9 Batch    9/269 - Train Accuracy: 0.9018, Validation Accuracy: 0.9020, Loss: 0.1369
Epoch   9 Batch   10/269 - Train Accuracy: 0.9062, Validation Accuracy: 0.8988, Loss: 0.1269
Epoch   9 Batch   11/269 - Train Accuracy: 0.9049, Validation Accuracy

Epoch   9 Batch   90/269 - Train Accuracy: 0.8926, Validation Accuracy: 0.9062, Loss: 0.1278
Epoch   9 Batch   91/269 - Train Accuracy: 0.9170, Validation Accuracy: 0.8987, Loss: 0.1157
Epoch   9 Batch   92/269 - Train Accuracy: 0.9229, Validation Accuracy: 0.8981, Loss: 0.1133
Epoch   9 Batch   93/269 - Train Accuracy: 0.9079, Validation Accuracy: 0.8973, Loss: 0.1171
Epoch   9 Batch   94/269 - Train Accuracy: 0.9012, Validation Accuracy: 0.9055, Loss: 0.1364
Epoch   9 Batch   95/269 - Train Accuracy: 0.9132, Validation Accuracy: 0.9041, Loss: 0.1137
Epoch   9 Batch   96/269 - Train Accuracy: 0.9039, Validation Accuracy: 0.9047, Loss: 0.1234
Epoch   9 Batch   97/269 - Train Accuracy: 0.8958, Validation Accuracy: 0.9001, Loss: 0.1245
Epoch   9 Batch   98/269 - Train Accuracy: 0.9036, Validation Accuracy: 0.8956, Loss: 0.1267
Epoch   9 Batch   99/269 - Train Accuracy: 0.9024, Validation Accuracy: 0.9031, Loss: 0.1294
Epoch   9 Batch  100/269 - Train Accuracy: 0.9165, Validation Accuracy

Epoch   9 Batch  179/269 - Train Accuracy: 0.9006, Validation Accuracy: 0.9093, Loss: 0.1154
Epoch   9 Batch  180/269 - Train Accuracy: 0.9145, Validation Accuracy: 0.9118, Loss: 0.1136
Epoch   9 Batch  181/269 - Train Accuracy: 0.8958, Validation Accuracy: 0.9141, Loss: 0.1211
Epoch   9 Batch  182/269 - Train Accuracy: 0.9129, Validation Accuracy: 0.9139, Loss: 0.1142
Epoch   9 Batch  183/269 - Train Accuracy: 0.9190, Validation Accuracy: 0.9158, Loss: 0.0948
Epoch   9 Batch  184/269 - Train Accuracy: 0.9171, Validation Accuracy: 0.9118, Loss: 0.1174
Epoch   9 Batch  185/269 - Train Accuracy: 0.9157, Validation Accuracy: 0.9077, Loss: 0.1184
Epoch   9 Batch  186/269 - Train Accuracy: 0.9190, Validation Accuracy: 0.9088, Loss: 0.1124
Epoch   9 Batch  187/269 - Train Accuracy: 0.9155, Validation Accuracy: 0.9102, Loss: 0.1081
Epoch   9 Batch  188/269 - Train Accuracy: 0.9286, Validation Accuracy: 0.9126, Loss: 0.1094
Epoch   9 Batch  189/269 - Train Accuracy: 0.9235, Validation Accuracy

Epoch  10 Batch    1/269 - Train Accuracy: 0.9064, Validation Accuracy: 0.9102, Loss: 0.1124
Epoch  10 Batch    2/269 - Train Accuracy: 0.9030, Validation Accuracy: 0.9159, Loss: 0.1231
Epoch  10 Batch    3/269 - Train Accuracy: 0.9194, Validation Accuracy: 0.9107, Loss: 0.1191
Epoch  10 Batch    4/269 - Train Accuracy: 0.9068, Validation Accuracy: 0.9134, Loss: 0.1228
Epoch  10 Batch    5/269 - Train Accuracy: 0.8869, Validation Accuracy: 0.8955, Loss: 0.1136
Epoch  10 Batch    6/269 - Train Accuracy: 0.9310, Validation Accuracy: 0.9145, Loss: 0.1112
Epoch  10 Batch    7/269 - Train Accuracy: 0.9118, Validation Accuracy: 0.9066, Loss: 0.1113
Epoch  10 Batch    8/269 - Train Accuracy: 0.9281, Validation Accuracy: 0.9188, Loss: 0.1181
Epoch  10 Batch    9/269 - Train Accuracy: 0.9038, Validation Accuracy: 0.9090, Loss: 0.1109
Epoch  10 Batch   10/269 - Train Accuracy: 0.9183, Validation Accuracy: 0.9134, Loss: 0.1170
Epoch  10 Batch   11/269 - Train Accuracy: 0.9182, Validation Accuracy

Epoch  10 Batch   90/269 - Train Accuracy: 0.9143, Validation Accuracy: 0.9242, Loss: 0.1095
Epoch  10 Batch   91/269 - Train Accuracy: 0.9273, Validation Accuracy: 0.9216, Loss: 0.0967
Epoch  10 Batch   92/269 - Train Accuracy: 0.9289, Validation Accuracy: 0.9181, Loss: 0.0984
Epoch  10 Batch   93/269 - Train Accuracy: 0.9281, Validation Accuracy: 0.9280, Loss: 0.0959
Epoch  10 Batch   94/269 - Train Accuracy: 0.9090, Validation Accuracy: 0.9236, Loss: 0.1168
Epoch  10 Batch   95/269 - Train Accuracy: 0.9103, Validation Accuracy: 0.9221, Loss: 0.1014
Epoch  10 Batch   96/269 - Train Accuracy: 0.9138, Validation Accuracy: 0.9201, Loss: 0.1036
Epoch  10 Batch   97/269 - Train Accuracy: 0.9097, Validation Accuracy: 0.9216, Loss: 0.1046
Epoch  10 Batch   98/269 - Train Accuracy: 0.9216, Validation Accuracy: 0.9206, Loss: 0.1030
Epoch  10 Batch   99/269 - Train Accuracy: 0.9233, Validation Accuracy: 0.9213, Loss: 0.1032
Epoch  10 Batch  100/269 - Train Accuracy: 0.9259, Validation Accuracy

Epoch  10 Batch  179/269 - Train Accuracy: 0.9066, Validation Accuracy: 0.9307, Loss: 0.0958
Epoch  10 Batch  180/269 - Train Accuracy: 0.9217, Validation Accuracy: 0.9263, Loss: 0.0933
Epoch  10 Batch  181/269 - Train Accuracy: 0.9136, Validation Accuracy: 0.9297, Loss: 0.1026
Epoch  10 Batch  182/269 - Train Accuracy: 0.9115, Validation Accuracy: 0.9229, Loss: 0.0967
Epoch  10 Batch  183/269 - Train Accuracy: 0.9276, Validation Accuracy: 0.9242, Loss: 0.0811
Epoch  10 Batch  184/269 - Train Accuracy: 0.9314, Validation Accuracy: 0.9272, Loss: 0.0943
Epoch  10 Batch  185/269 - Train Accuracy: 0.9171, Validation Accuracy: 0.9267, Loss: 0.1000
Epoch  10 Batch  186/269 - Train Accuracy: 0.9270, Validation Accuracy: 0.9296, Loss: 0.0966
Epoch  10 Batch  187/269 - Train Accuracy: 0.9320, Validation Accuracy: 0.9300, Loss: 0.0913
Epoch  10 Batch  188/269 - Train Accuracy: 0.9346, Validation Accuracy: 0.9244, Loss: 0.0979
Epoch  10 Batch  189/269 - Train Accuracy: 0.9260, Validation Accuracy

Epoch  11 Batch    1/269 - Train Accuracy: 0.9071, Validation Accuracy: 0.9253, Loss: 0.0947
Epoch  11 Batch    2/269 - Train Accuracy: 0.9174, Validation Accuracy: 0.9244, Loss: 0.0951
Epoch  11 Batch    3/269 - Train Accuracy: 0.9345, Validation Accuracy: 0.9243, Loss: 0.0973
Epoch  11 Batch    4/269 - Train Accuracy: 0.9138, Validation Accuracy: 0.9216, Loss: 0.0957
Epoch  11 Batch    5/269 - Train Accuracy: 0.9194, Validation Accuracy: 0.9164, Loss: 0.0956
Epoch  11 Batch    6/269 - Train Accuracy: 0.9368, Validation Accuracy: 0.9210, Loss: 0.0838
Epoch  11 Batch    7/269 - Train Accuracy: 0.9212, Validation Accuracy: 0.9179, Loss: 0.0916
Epoch  11 Batch    8/269 - Train Accuracy: 0.9403, Validation Accuracy: 0.9183, Loss: 0.0960
Epoch  11 Batch    9/269 - Train Accuracy: 0.9212, Validation Accuracy: 0.9244, Loss: 0.0934
Epoch  11 Batch   10/269 - Train Accuracy: 0.9283, Validation Accuracy: 0.9243, Loss: 0.0896
Epoch  11 Batch   11/269 - Train Accuracy: 0.9244, Validation Accuracy

Epoch  11 Batch   90/269 - Train Accuracy: 0.9199, Validation Accuracy: 0.9269, Loss: 0.1009
Epoch  11 Batch   91/269 - Train Accuracy: 0.9330, Validation Accuracy: 0.9292, Loss: 0.0881
Epoch  11 Batch   92/269 - Train Accuracy: 0.9368, Validation Accuracy: 0.9282, Loss: 0.0861
Epoch  11 Batch   93/269 - Train Accuracy: 0.9329, Validation Accuracy: 0.9291, Loss: 0.0862
Epoch  11 Batch   94/269 - Train Accuracy: 0.9230, Validation Accuracy: 0.9317, Loss: 0.1087
Epoch  11 Batch   95/269 - Train Accuracy: 0.9191, Validation Accuracy: 0.9253, Loss: 0.0941
Epoch  11 Batch   96/269 - Train Accuracy: 0.9148, Validation Accuracy: 0.9372, Loss: 0.0922
Epoch  11 Batch   97/269 - Train Accuracy: 0.9238, Validation Accuracy: 0.9310, Loss: 0.0956
Epoch  11 Batch   98/269 - Train Accuracy: 0.9357, Validation Accuracy: 0.9279, Loss: 0.0914
Epoch  11 Batch   99/269 - Train Accuracy: 0.9340, Validation Accuracy: 0.9231, Loss: 0.0891
Epoch  11 Batch  100/269 - Train Accuracy: 0.9327, Validation Accuracy

Epoch  11 Batch  179/269 - Train Accuracy: 0.9116, Validation Accuracy: 0.9305, Loss: 0.0836
Epoch  11 Batch  180/269 - Train Accuracy: 0.9264, Validation Accuracy: 0.9272, Loss: 0.0841
Epoch  11 Batch  181/269 - Train Accuracy: 0.9169, Validation Accuracy: 0.9330, Loss: 0.0913
Epoch  11 Batch  182/269 - Train Accuracy: 0.9178, Validation Accuracy: 0.9326, Loss: 0.0863
Epoch  11 Batch  183/269 - Train Accuracy: 0.9333, Validation Accuracy: 0.9316, Loss: 0.0775
Epoch  11 Batch  184/269 - Train Accuracy: 0.9339, Validation Accuracy: 0.9346, Loss: 0.0851
Epoch  11 Batch  185/269 - Train Accuracy: 0.9347, Validation Accuracy: 0.9278, Loss: 0.0886
Epoch  11 Batch  186/269 - Train Accuracy: 0.9266, Validation Accuracy: 0.9290, Loss: 0.0833
Epoch  11 Batch  187/269 - Train Accuracy: 0.9302, Validation Accuracy: 0.9264, Loss: 0.0826
Epoch  11 Batch  188/269 - Train Accuracy: 0.9348, Validation Accuracy: 0.9326, Loss: 0.0860
Epoch  11 Batch  189/269 - Train Accuracy: 0.9284, Validation Accuracy

Epoch  12 Batch    1/269 - Train Accuracy: 0.9187, Validation Accuracy: 0.9390, Loss: 0.0831
Epoch  12 Batch    2/269 - Train Accuracy: 0.9242, Validation Accuracy: 0.9391, Loss: 0.0849
Epoch  12 Batch    3/269 - Train Accuracy: 0.9411, Validation Accuracy: 0.9418, Loss: 0.0853
Epoch  12 Batch    4/269 - Train Accuracy: 0.9176, Validation Accuracy: 0.9377, Loss: 0.0819
Epoch  12 Batch    5/269 - Train Accuracy: 0.9258, Validation Accuracy: 0.9329, Loss: 0.0851
Epoch  12 Batch    6/269 - Train Accuracy: 0.9422, Validation Accuracy: 0.9333, Loss: 0.0782
Epoch  12 Batch    7/269 - Train Accuracy: 0.9326, Validation Accuracy: 0.9364, Loss: 0.0797
Epoch  12 Batch    8/269 - Train Accuracy: 0.9414, Validation Accuracy: 0.9367, Loss: 0.0921
Epoch  12 Batch    9/269 - Train Accuracy: 0.9296, Validation Accuracy: 0.9324, Loss: 0.0812
Epoch  12 Batch   10/269 - Train Accuracy: 0.9371, Validation Accuracy: 0.9339, Loss: 0.0780
Epoch  12 Batch   11/269 - Train Accuracy: 0.9338, Validation Accuracy

Epoch  12 Batch   90/269 - Train Accuracy: 0.9295, Validation Accuracy: 0.9360, Loss: 0.0846
Epoch  12 Batch   91/269 - Train Accuracy: 0.9345, Validation Accuracy: 0.9373, Loss: 0.0750
Epoch  12 Batch   92/269 - Train Accuracy: 0.9453, Validation Accuracy: 0.9362, Loss: 0.0705
Epoch  12 Batch   93/269 - Train Accuracy: 0.9369, Validation Accuracy: 0.9403, Loss: 0.0744
Epoch  12 Batch   94/269 - Train Accuracy: 0.9213, Validation Accuracy: 0.9407, Loss: 0.0928
Epoch  12 Batch   95/269 - Train Accuracy: 0.9294, Validation Accuracy: 0.9420, Loss: 0.0839
Epoch  12 Batch   96/269 - Train Accuracy: 0.9243, Validation Accuracy: 0.9314, Loss: 0.0771
Epoch  12 Batch   97/269 - Train Accuracy: 0.9300, Validation Accuracy: 0.9355, Loss: 0.0835
Epoch  12 Batch   98/269 - Train Accuracy: 0.9347, Validation Accuracy: 0.9362, Loss: 0.0792
Epoch  12 Batch   99/269 - Train Accuracy: 0.9331, Validation Accuracy: 0.9307, Loss: 0.0788
Epoch  12 Batch  100/269 - Train Accuracy: 0.9403, Validation Accuracy

Epoch  12 Batch  179/269 - Train Accuracy: 0.9219, Validation Accuracy: 0.9347, Loss: 0.0732
Epoch  12 Batch  180/269 - Train Accuracy: 0.9432, Validation Accuracy: 0.9372, Loss: 0.0718
Epoch  12 Batch  181/269 - Train Accuracy: 0.9224, Validation Accuracy: 0.9347, Loss: 0.0864
Epoch  12 Batch  182/269 - Train Accuracy: 0.9260, Validation Accuracy: 0.9367, Loss: 0.0811
Epoch  12 Batch  183/269 - Train Accuracy: 0.9395, Validation Accuracy: 0.9370, Loss: 0.0659
Epoch  12 Batch  184/269 - Train Accuracy: 0.9384, Validation Accuracy: 0.9335, Loss: 0.0763
Epoch  12 Batch  185/269 - Train Accuracy: 0.9404, Validation Accuracy: 0.9349, Loss: 0.0755
Epoch  12 Batch  186/269 - Train Accuracy: 0.9335, Validation Accuracy: 0.9386, Loss: 0.0716
Epoch  12 Batch  187/269 - Train Accuracy: 0.9383, Validation Accuracy: 0.9395, Loss: 0.0750
Epoch  12 Batch  188/269 - Train Accuracy: 0.9454, Validation Accuracy: 0.9436, Loss: 0.0720
Epoch  12 Batch  189/269 - Train Accuracy: 0.9390, Validation Accuracy

Epoch  13 Batch    1/269 - Train Accuracy: 0.9276, Validation Accuracy: 0.9419, Loss: 0.0770
Epoch  13 Batch    2/269 - Train Accuracy: 0.9315, Validation Accuracy: 0.9463, Loss: 0.0815
Epoch  13 Batch    3/269 - Train Accuracy: 0.9479, Validation Accuracy: 0.9445, Loss: 0.0785
Epoch  13 Batch    4/269 - Train Accuracy: 0.9262, Validation Accuracy: 0.9426, Loss: 0.0763
Epoch  13 Batch    5/269 - Train Accuracy: 0.9349, Validation Accuracy: 0.9403, Loss: 0.0761
Epoch  13 Batch    6/269 - Train Accuracy: 0.9498, Validation Accuracy: 0.9438, Loss: 0.0694
Epoch  13 Batch    7/269 - Train Accuracy: 0.9350, Validation Accuracy: 0.9401, Loss: 0.0693
Epoch  13 Batch    8/269 - Train Accuracy: 0.9523, Validation Accuracy: 0.9462, Loss: 0.0815
Epoch  13 Batch    9/269 - Train Accuracy: 0.9380, Validation Accuracy: 0.9480, Loss: 0.0768
Epoch  13 Batch   10/269 - Train Accuracy: 0.9435, Validation Accuracy: 0.9437, Loss: 0.0695
Epoch  13 Batch   11/269 - Train Accuracy: 0.9247, Validation Accuracy

Epoch  13 Batch   90/269 - Train Accuracy: 0.9390, Validation Accuracy: 0.9459, Loss: 0.0815
Epoch  13 Batch   91/269 - Train Accuracy: 0.9408, Validation Accuracy: 0.9379, Loss: 0.0677
Epoch  13 Batch   92/269 - Train Accuracy: 0.9501, Validation Accuracy: 0.9413, Loss: 0.0646
Epoch  13 Batch   93/269 - Train Accuracy: 0.9382, Validation Accuracy: 0.9480, Loss: 0.0682
Epoch  13 Batch   94/269 - Train Accuracy: 0.9364, Validation Accuracy: 0.9395, Loss: 0.0823
Epoch  13 Batch   95/269 - Train Accuracy: 0.9391, Validation Accuracy: 0.9416, Loss: 0.0731
Epoch  13 Batch   96/269 - Train Accuracy: 0.9237, Validation Accuracy: 0.9425, Loss: 0.0781
Epoch  13 Batch   97/269 - Train Accuracy: 0.9256, Validation Accuracy: 0.9326, Loss: 0.0764
Epoch  13 Batch   98/269 - Train Accuracy: 0.9330, Validation Accuracy: 0.9375, Loss: 0.0722
Epoch  13 Batch   99/269 - Train Accuracy: 0.9353, Validation Accuracy: 0.9455, Loss: 0.0730
Epoch  13 Batch  100/269 - Train Accuracy: 0.9440, Validation Accuracy

Epoch  13 Batch  179/269 - Train Accuracy: 0.9274, Validation Accuracy: 0.9481, Loss: 0.0670
Epoch  13 Batch  180/269 - Train Accuracy: 0.9430, Validation Accuracy: 0.9430, Loss: 0.0622
Epoch  13 Batch  181/269 - Train Accuracy: 0.9426, Validation Accuracy: 0.9439, Loss: 0.0756
Epoch  13 Batch  182/269 - Train Accuracy: 0.9410, Validation Accuracy: 0.9427, Loss: 0.0682
Epoch  13 Batch  183/269 - Train Accuracy: 0.9403, Validation Accuracy: 0.9426, Loss: 0.0586
Epoch  13 Batch  184/269 - Train Accuracy: 0.9363, Validation Accuracy: 0.9429, Loss: 0.0697
Epoch  13 Batch  185/269 - Train Accuracy: 0.9461, Validation Accuracy: 0.9426, Loss: 0.0644
Epoch  13 Batch  186/269 - Train Accuracy: 0.9403, Validation Accuracy: 0.9459, Loss: 0.0622
Epoch  13 Batch  187/269 - Train Accuracy: 0.9494, Validation Accuracy: 0.9443, Loss: 0.0634
Epoch  13 Batch  188/269 - Train Accuracy: 0.9472, Validation Accuracy: 0.9437, Loss: 0.0670
Epoch  13 Batch  189/269 - Train Accuracy: 0.9386, Validation Accuracy

Epoch  14 Batch    1/269 - Train Accuracy: 0.9429, Validation Accuracy: 0.9520, Loss: 0.0727
Epoch  14 Batch    2/269 - Train Accuracy: 0.9368, Validation Accuracy: 0.9500, Loss: 0.0722
Epoch  14 Batch    3/269 - Train Accuracy: 0.9491, Validation Accuracy: 0.9516, Loss: 0.0668
Epoch  14 Batch    4/269 - Train Accuracy: 0.9279, Validation Accuracy: 0.9473, Loss: 0.0732
Epoch  14 Batch    5/269 - Train Accuracy: 0.9354, Validation Accuracy: 0.9342, Loss: 0.0691
Epoch  14 Batch    6/269 - Train Accuracy: 0.9395, Validation Accuracy: 0.9328, Loss: 0.0620
Epoch  14 Batch    7/269 - Train Accuracy: 0.9417, Validation Accuracy: 0.9474, Loss: 0.0652
Epoch  14 Batch    8/269 - Train Accuracy: 0.9573, Validation Accuracy: 0.9466, Loss: 0.0684
Epoch  14 Batch    9/269 - Train Accuracy: 0.9386, Validation Accuracy: 0.9464, Loss: 0.0757
Epoch  14 Batch   10/269 - Train Accuracy: 0.9493, Validation Accuracy: 0.9513, Loss: 0.0644
Epoch  14 Batch   11/269 - Train Accuracy: 0.9411, Validation Accuracy

Epoch  14 Batch   90/269 - Train Accuracy: 0.9447, Validation Accuracy: 0.9482, Loss: 0.0664
Epoch  14 Batch   91/269 - Train Accuracy: 0.9496, Validation Accuracy: 0.9466, Loss: 0.0591
Epoch  14 Batch   92/269 - Train Accuracy: 0.9568, Validation Accuracy: 0.9449, Loss: 0.0575
Epoch  14 Batch   93/269 - Train Accuracy: 0.9479, Validation Accuracy: 0.9471, Loss: 0.0590
Epoch  14 Batch   94/269 - Train Accuracy: 0.9340, Validation Accuracy: 0.9515, Loss: 0.0708
Epoch  14 Batch   95/269 - Train Accuracy: 0.9422, Validation Accuracy: 0.9506, Loss: 0.0647
Epoch  14 Batch   96/269 - Train Accuracy: 0.9308, Validation Accuracy: 0.9466, Loss: 0.0669
Epoch  14 Batch   97/269 - Train Accuracy: 0.9347, Validation Accuracy: 0.9443, Loss: 0.0654
Epoch  14 Batch   98/269 - Train Accuracy: 0.9394, Validation Accuracy: 0.9447, Loss: 0.0682
Epoch  14 Batch   99/269 - Train Accuracy: 0.9439, Validation Accuracy: 0.9478, Loss: 0.0619
Epoch  14 Batch  100/269 - Train Accuracy: 0.9481, Validation Accuracy

Epoch  14 Batch  179/269 - Train Accuracy: 0.9314, Validation Accuracy: 0.9533, Loss: 0.0592
Epoch  14 Batch  180/269 - Train Accuracy: 0.9541, Validation Accuracy: 0.9524, Loss: 0.0611
Epoch  14 Batch  181/269 - Train Accuracy: 0.9448, Validation Accuracy: 0.9497, Loss: 0.0650
Epoch  14 Batch  182/269 - Train Accuracy: 0.9443, Validation Accuracy: 0.9419, Loss: 0.0613
Epoch  14 Batch  183/269 - Train Accuracy: 0.9515, Validation Accuracy: 0.9442, Loss: 0.0533
Epoch  14 Batch  184/269 - Train Accuracy: 0.9534, Validation Accuracy: 0.9517, Loss: 0.0612
Epoch  14 Batch  185/269 - Train Accuracy: 0.9547, Validation Accuracy: 0.9487, Loss: 0.0592
Epoch  14 Batch  186/269 - Train Accuracy: 0.9376, Validation Accuracy: 0.9508, Loss: 0.0591
Epoch  14 Batch  187/269 - Train Accuracy: 0.9465, Validation Accuracy: 0.9515, Loss: 0.0633
Epoch  14 Batch  188/269 - Train Accuracy: 0.9489, Validation Accuracy: 0.9535, Loss: 0.0618
Epoch  14 Batch  189/269 - Train Accuracy: 0.9470, Validation Accuracy

Model Trained and Saved


### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [21]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [22]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [23]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    word_ids = []
    for word in sentence.lower().split():
        if word in vocab_to_int:
            word_ids.append(vocab_to_int[word])
        else:
            word_ids.append(vocab_to_int['<UNK>'])
    return word_ids


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [24]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [61, 91, 166, 225, 107, 127, 206]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [237, 167, 156, 122, 13, 218, 66, 88, 1]
  French Words: il a vu un vieux camion jaune . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.